server/src/main/java/org/elasticsearch/common/bytes/BytesReferenceStreamInput.java

```diff
@@ -174,6 +175,53 @@ public long readVLong() throws IOException {
         }
     }
 
+    @Override
+    protected String readString(int charCount) throws IOException {
+        // use fast path when we have enough bytes to always safely read the requested number of chars or when we are on the last slice
+        // anyway and will just be reading from this single array
+        final int currentSliceLength = slice.length;
+        if (charCount * 3 <= currentSliceLength - sliceIndex || sliceStartOffset + currentSliceLength == bytesReference.length()) {
+            final CharsRef charsRef = getCharsRef(charCount);
+            final char[] charBuffer = charsRef.chars;
+            final byte[] byteBuffer = slice.bytes;
+            final int offset = slice.offset;
+            int offsetByteArray = sliceIndex + offset;
+            for (int charsOffset = 0; charsOffset < charCount; charsOffset++) {
```

Comment:
If we are going to reimplement utf8 decoding, we should have lots of tests for it...

Can we instead use CharsetDecoder? We should just need to wrap the byte[] in ByteBuffer.

Reply:
I way cut back on the complexity of the implementation here now and just made the fast-path against arrays work for ascii (the way it's implemented now that's a much bigger win in the real than the first version ever was anyway).
I'm not sure we can tune charset decoder easily while retaining the current performance (you'd have a very large `final` method on the charset parent that won't inline so well and some object creation for the decode result). Even the JDK's `java.io.DataInputStream#readUTF(java.io.DataInput)` (which is essentially the same as our encoding except that they encode byte-size and not char-count in the prefix) isn't using a charset and instead relies on a hand-crafted loop.

That said, we do have massive test coverage on this code IMO. Tests like e.g. `org.elasticsearch.common.io.stream.BytesStreamsTests#testWriteRandomStrings` give us 100% code coverage (which essentially means 100% case coverage too here I think since it's really just about getting all the size+offset combinations right).

Reply:
My understanding of `BytesStreamsTests#testWriteRandomStrings` is that it will almost always yield non-ASCII strings, and will use 16kiB pages each backed by a 16kiB `byte[]` so the offset is zero. I'd like to see more cases involving all/mostly ASCII characters and much weirder page boundaries and nonzero offsets.

Reply:
> Even the JDK's java.io.DataInputStream#readUTF(java.io.DataInput) ... isn't using a charset and instead relies on a hand-crafted loop.

DataInputStream has been around since 1.0, while CharsetDecoder was introduced in 1.4. I'm not sure it can be implied the reason a hand-crafted loop exists is for performance reasons. The UTF8 decoder has a fastpath for ascii. It first attempts to read as many characters as it can using ascii, then bails on end or non-ascii and decodes the rest using a slower loop if necessary. This should be easy to swap into your branch and test?

My larger concern on the testing is that we must maintain this complexity. UTF8 is not simple, and we can spend a lot of time trying to optimize our own implementation, and then have tests for it. But the JDK already must do this itself, and they have spent time optimizing CharsetDecoder, so let's use the mechanism the JDK provides.

Reply:
The problem is that the JDK optimises for the general case. I bet their version with the internal ascii fast-path works out great for longer strings and it offers a lot more functionality than we need (ignoring broken byte sequences etc.). But for us things in profiling generally already look like this:

<img width="841" alt="image" src="https://user-images.githubusercontent.com/6490959/161150061-257bd251-9f9e-4759-b066-550bf4f7f4a5.png">

We have a gigantic method going through all the rare corner cases so nothing inlines and it itself isn't inlined. On top of that we needlessly copy bytes because the interface is a stream even though what we actually have is one or more byte arrays.  If we go for the JDKs solution here, we'll just get even more indirection.
The average string we're dealing with in the very hot methods is probably ~25 chars in size. A loop of 25 iterations over a byte array without any further method calls on the fast paths will probably always beat a very general solution that has to wrap the byte array in a buffer, do the same for the char array and do a bunch of method calls afterwards.
This might be different if we were decoding strings of 4k chars or so, but outside of exceptions messages we generally don't do that much.

Reply:
> A loop of 25 iterations over a byte array without any further method calls on the fast paths will probably always beat a very general solution that has to wrap the byte array in a buffer, do the same for the char array and do a bunch of method calls afterwards.

Maybe, but there are a bunch of assumptions there about things the way they are now. The fact that we would need to wrap the slice is because we use BytesRef, but we could instead use ByteBuffer as our view to the slice for the entire BytesReferenceStreamInput. The CharBuffer doesn't need to be wrapped, we could store that in the thread local instead of the raw char[].

With what I'm suggesting we would still have a fast path. That is, the first call to CharsetDecoder completely decodes the string. And that decoder itself has a very fast fast-path, the ascii decode I mentioned is an intrinsic, which we are not getting by doing our own decoding (see https://cl4es.github.io/2021/02/23/Faster-Charset-Decoding.html). It's exactly that type of optimization by the JDK which I think is being underestimated here.